### PR TITLE
Update for SIP5 build

### PR DESCRIPTION
**Pull request by ustsv**
Update for SIP5 build:
- find SIP5 executable and version
- PyQt5.sip
- gen and include sip.h
- deprecated function in .sip

### CODE REVIEW

**Comment by Ghostkeeper** [comment]
There are a number of problems with this pull request:
* It doesn't seem to work. At least, the automated test is failing with that it can't find sip.h. We'd also need to update the build environment to have this SIP version.
* It makes this repository depend on PyQt5. libSavitar doesn't have anything to do with Qt so that's an unreasonable dependency.
* I don't really see a reason for us to update. We're not using anything in SIP 5.

For these reasons I'll reject this pull request, sorry.

**Comment by ustsv** [comment]
Ok, leave for the future.
In MSYS2 / MinGW64 this option is built and works. Sip5 replace sip(4) and sip python  module as PyQt5.sip.
Test git version PKGBUILD https://bitbucket.org/ustsv/testpkg_msys2/src/master/

**Comment by StefanBruens** [comment]
@Ghostkeeper - I think there is some confusion between the PyQt5 module name (which is just a namespace) and PyQt5 itself. Using 'PyQt5.sip' as namespace does not add any build or runtime dependency.

Having an non-namespaced `sip` has caused issues for distribution packages in the past (Qt4 vs Qt5) and with Qt6 around the corner this will likely happen again. Being able to use a fully-qualified module name (thats how Riverbank calls it) definitely helps downstream distros.

How about adding a CMake option to specify the full module name, defaulting to "sip"? Note, most downstreams already default to 'PyQt5.sip' today, patching libArcus, libSavitar and python-libnest2d.

For reference, see https://github.com/Ultimaker/libArcus/issues/76

File: cmake/FindSIP.cmake [hunk 1]
```diff
@@ -68,16 +68,26 @@ endif()
...
 )
 
+execute_process(
+  COMMAND sip-module --sip-h --target-dir ${CMAKE_CURRENT_BINARY_DIR} PyQt5.sip
+  RESULT_VARIABLE _process_status
+  OUTPUT_VARIABLE _process_output
+  OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+if(NOT ${_process_status} EQUAL 0)
+  message(FATAL_ERROR "Failed to get sip.h. Error: ${_process_output}")
```
Comment (StefanBruens) [inline cmake/FindSIP.cmake:82]: Should be "Failed to **generate** sip.h"

File: cmake/FindSIP.cmake [hunk 2]
```diff
@@ -68,16 +68,26 @@ endif()
 
 get_filename_component(_python_binary_path ${Python3_EXECUTABLE} DIRECTORY)
 
-find_program(SIP_EXECUTABLE sip
+find_program(SIP_EXECUTABLE sip5
```
Comment (StefanBruens) [inline cmake/FindSIP.cmake:71]: should be 
```
find_program(SIP_EXECUTABLE
    NAMES sip sip4 sip5
    ...
```

File: cmake/FindSIP.cmake [hunk 3]
```diff
@@ -68,16 +68,26 @@ endif()
...
     HINTS ${CMAKE_PREFIX_PATH}/bin ${CMAKE_INSTALL_PATH}/bin ${_python_binary_path} ${Python3_SITELIB}/PyQt5
 )
 
+execute_process(
+  COMMAND sip-module --sip-h --target-dir ${CMAKE_CURRENT_BINARY_DIR} PyQt5.sip
```
Comment (StefanBruens) [inline cmake/FindSIP.cmake:76]: The module name should be an CMAKE_OPTION, defaulting to 'sip' for legacy compat.

File: python/Types.sip [hunk 4]
```diff
@@ -94,7 +94,7 @@
         for (int i = 0; i < (int)sipCpp->size(); ++i)
         {
             SceneNode *cpp = new SceneNode(*sipCpp->at(i));
-            PyObject *pobj = sipConvertFromInstance(cpp, sipClass_SceneNode, sipTransferObj);
+            PyObject *pobj = sipConvertFromType(cpp, sipType_SceneNode, sipTransferObj);
```
Comment (StefanBruens) [inline python/Types.sip:97]: @Ghostkeeper 
This is just an update from Sip ABI 11 (Sip <= 4.18.x) to Sip ABI 12 (Sip >= 4.19.0). As Sip 4.19.2 is the minimum required version, this should be correct anyway (and removes deprecated code).

Comment (StefanBruens) [reply python/Types.sip:97]: Its even much older, sipConvertFromInstance is deprecated since 4.8.0!

File: cmake/FindSIP.cmake (hunk 3 above)
Comment (bnavigator) [reply cmake/FindSIP.cmake:76]: Moreover, you should not rebuild the existing PyQt5.sip module and conflict with an already installed one. If you really want to build your own (why?), use a unique name.

File: cmake/FindSIP.cmake [hunk 5]
```diff
@@ -68,16 +68,26 @@ endif()
...
+endif()
+
 find_path(SIP_INCLUDE_DIRS sip.h
-    HINTS ${CMAKE_PREFIX_PATH}/include ${CMAKE_INSTALL_PATH}/include ${Python3_INCLUDE_DIRS} ${Python3_SITELIB}/PyQt5
+    HINTS ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_PREFIX_PATH}/include ${CMAKE_INSTALL_PATH}/include ${Python3_INCLUDE_DIRS} ${Python3_SITELIB}/PyQt5
 )
 
 execute_process(
-    COMMAND ${Python3_EXECUTABLE} -c "import sip; print(sip.SIP_VERSION_STR)"
+    COMMAND ${Python3_EXECUTABLE} -c "import sipbuild; print(sipbuild.builder.SIP_VERSION_STR)"
```
Comment (bnavigator) [inline cmake/FindSIP.cmake:90]: What's the advantage over `sip-build -V` here?

**Comment by bnavigator** [comment]
> Using 'PyQt5.sip' as namespace does not add any build or runtime dependency.

This is not correct. If you don't build it inline, you will always have a runtime dependency on the sip module.

Build patched to use -n PyQt5.sip, but without PyQt5.sip installed at runtime:

```
greiner@greinerT450s:~/src/osc/home:bnavigator:branches:science/libArcus/ > osc shell --login-as-root
greinerT450s:~ # ls -l /usr/lib64/python3.8/site-packages/PyQt5
ls: cannot access '/usr/lib64/python3.8/site-packages/PyQt5': No such file or directory
greinerT450s:~ # rpm -i /home/abuild/rpmbuild/RPMS/x86_64/libArcus3-4.8.0-0.x86_64.rpm /home/abuild/rpmbuild/RPMS/x86_64/python3-Arcus-4.8.0-0.x86_64.rpm
greinerT450s:~ # python3
Python 3.8.6 (default, Nov 09 2020, 12:09:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import Arcus
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'PyQt5'
```

With PyQt5.sip available:
```
greiner@greinerT450s:~/src/osc/home:bnavigator:branches:science/libArcus/ > osc shell --login-as-root
greinerT450s:~ # ls -l /usr/lib64/python3.8/site-packages/PyQt5
total 160
-rwxr-xr-x 1 root root 161464 Nov 12 22:01 sip.cpython-38-x86_64-linux-gnu.so
greinerT450s:~ # rpm -i /home/abuild/rpmbuild/RPMS/x86_64/libArcus3-4.8.0-0.x86_64.rpm /home/abuild/rpmbuild/RPMS/x86_64/python3-Arcus-4.8.0-0.x86_64.rpm
greinerT450s:~ # python3
Python 3.8.6 (default, Nov 09 2020, 12:09:06) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import Arcus
>>>   
greinerT450s:~ # 
```

This is true for libArcus, libSavitar and pynest2d. Note that PyQt5 itself is not installed, neither is Qt. The only file in site-packages/PyQt5 is the sip module.


If you want to use those modules with cura (the primary use case, isn't it), which depends on PyQt5, you are bound to use the same module as cura pulls in through PyQt5. That is the reason why most distros patch the modules to use the PyQt5.sip namespace, as already mentioned: Ultimaker/libArcus#76

**Comment by Ghostkeeper** [comment]
A few things you need to understand, as expectation management here. We develop Cura for work, meaning that we each get a limited number of hours and a sprint to complete. There is some leeway in this, but there's a lot to do every day. We can (and do) also work on Cura-related things in our spare time. But if that is part of our main code base (not a plug-in) then it needs to be reviewed and tested in many cases. Changing a dependency of 3 of our libraries which is also central to the most difficult part of building Cura for Windows is certainly not something I'd do on my own in my spare time. All other developers will be impacted, our build system needs to be updated on 4 servers, and we'd need to do an integration test. So that leaves Ultimaker's time to implement this.

Ultimaker has its own distribution through the .AppImage file for Linux (~0.7% of Cura's users atm). If it can be made easy for people to create a package "native" for certain distributions, then we'd be happy to oblige, but if it impacts our planning then Ultimaker **will not** make this change unless there is a direct benefit for our own distribution method.

As a result:

* Changing the Sip version is a costly change. This **will not happen** unless there is a demonstrable benefit for our own version of Cura.
* Changing `sipConvertFromInstance` into `sipConvertFromType` is okay with me as long as `sipConvertFromType` works the same way with our current Sip version. If it's a drop-in replacement for our usage we can probably get that through pretty fast.
* Changing the CMake such that it works with both Sip versions is somewhat iffy because reasonably a user can then expect us to maintain compatibility with Sip 5. This may not really be a problem through documentation. In general, Sip 5 is not something we'll be able to test.

**Pull request closed without merging**